warehouse/admin/views/users.py

```diff
@@ -182,6 +185,51 @@ def user_add_email(request):
     return HTTPSeeOther(request.route_path("admin.user.detail", user_id=user.id))
 
 
+@view_config(
+    route_name="admin.user.reset_2fa",
+    require_methods=["POST"],
+    permission="admin",
+    uses_session=True,
+    require_csrf=True,
+)
+def user_reset_2fa(request):
+    user = request.db.query(User).get(request.matchdict["user_id"])
+    form = UserForm(request.POST if request.method == "POST" else None, user)
+    gh_link = form.data["name"]
+
+    if form.validate():
+        valid_sites = set()
+        if len(user.projects) > 0:
+            options = {}
+            for i, project in enumerate([p for p in user.projects if len(p.releases)>0]):
+                options[i] = project
+                for k, v in project.releases[0].urls.items():
+                    try:
+                        status_code = requests.get(v).status_code
+                    except:
+                        status_code = 404
+
+                    if status_code == 200:
+                        valid_sites.add(project.name)
+
+            if valid_sites:
+                choice = random.choice(list(valid_sites))
```

Comment:
We probably don't want this to be random, but rather leave it up to the admin to choose. That probably means we need a two-step process, one that generates a list of URLs and another that takes a selection and sends the email.

Reply:
I suppose that would work.

I don't know if the long term goal would be able to have the user be able to do reset it themselves. In that case, the system would be auto-picking the projects, right? I figured, if we are headed that way, maybe we should include that now.

Reply:
I also wasn't sure if we should store the "token" (and probably selected project) used to validate the individual does have access to the code repository.

The reason is, I'm not sure how the warehouse admin would know what was sent to user, to have the admin validate everything matches.

Reply:
The long term goal is to keep a PyPI admin in the loop, just make it easier on that human -- this won't ever be self-serve.

Yes, the token should either be stored somewhere, made more deterministic, or just turned into something else -- I don't quite see the benefit to this being a random string. The goal is to prove that the person we're interacting with has control over the account/repository in question, so asking them to do something like push a commit to a predictable branch like `pypi-account-recovery-YYYY-MM-DD` with a commit message like:

> I (pypi.org/user/<username>) have requested account recovery on PyPI, and this commit signifies that I am in control of this corresponding account.

should be just as meaningful as a random string -- it just needs to be some activity that someone who is being compromised on PyPI would be unlikely to do on a corresponding account.

Reply:
Also summoning @ewdurbin for their thoughts here.

Reply:
I'm wrong here - we do want some sort of random token, that is sent via the email, to demonstrate ownership of the email account as well.

We should store this as well as the choice that's being made here, so when we verify we can refer back to this. This should probably be stored in a new model/table that represents a reset request.

Reply:
Indeed, a random secret allows us to validate that the actor who publishes a verification branch is the same as the one who controls the primary email address.

We'd need to store the verification code as well as the project(s) selected for validation. To proceed a single project is fine, but in some cases we may want someone to demonstrate ownership across certain boundaries (GitHub orgs or hosting services).

Reply:
Agreed, an administrator should review and select what project will be chosen for validation. Some key criteria we might look for which would be good to surface in the select view:

- Number of other maintainers/owners
- Is project marked as [critical](https://github.com/pypi/warehouse/blob/main/warehouse/packaging/models.py#L157)?
- Number of releases
- Most recent release
- # of valid/invalid URLs